Acquire read holds after the timestamp selection of a REFRESH MV

ggevay opened the pull request:
This fixes https://github.com/MaterializeInc/materialize/issues/24288 as discussed with @mjibson [here](https://materializeinc.slack.com/archives/C02FWJ94HME/p1704989974860729?thread_ts=1704982981.659269&cid=C02FWJ94HME).

### Motivation

  * This PR fixes a recognized bug: https://github.com/MaterializeInc/materialize/issues/24288

### Tips for reviewer

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

mjibson commented on src/adapter/src/coord/read_policy.rs:
```diff
@@ -419,6 +420,23 @@ impl crate::coord::Coordinator {
+    ) {
+        let read_holds = self.acquire_read_holds(time, id_bundle);
+        self.txn_read_holds
+            .insert(session.conn_id().clone(), read_holds);
```
The insert could delete some existing read holds. This should instead do something like `.entry(connid).or_insert_with(|| ReadHolds::new()).extend(read_holds)`

ggevay replied:
Oh, yes, sorry!

Actually, I copied this line from [here](https://github.com/MaterializeInc/materialize/blob/d6a3c785559d40564f9b70fb8536a43ac1e8cdc0/src/adapter/src/coord/sequencer/inner/peek.rs#L694). So then this is probably a bug also on `main`?

I've fixed both occurrences now.

mjibson commented on src/adapter/src/coord/command_handler.rs:
```diff
@@ -774,10 +785,17 @@ impl Coordinator {
+        &mut self,
         cmvs: &CreateMaterializedViewStatement<Aug>,
         resolved_ids: &ResolvedIds,
+        acquire_read_holds: bool,
```
Instead of a `bool` function argument, what about `acquire_read_holds_for: Option<&mut Session>`? That way the mut session is only passed if needed, and also serves as the flag to auto cleanup the read hold?

ggevay replied:
Good idea!

mjibson commented on src/adapter/src/coord/command_handler.rs:
```diff
@@ -821,11 +840,42 @@ impl Coordinator {
+    where
+        I: IntoIterator<Item = &'a GlobalId>,
+    {
+        let ids = ids.into_iter().collect_vec();
```
Low preference thought, ignore if you like: if this has to be a vec, seems reasonable to require the caller pass a vec?

ggevay replied:
Ok, I don't really have a preference, so I've changed it.

mjibson commented on src/adapter/src/coord/command_handler.rs:
```diff
@@ -821,11 +840,42 @@ impl Coordinator {
+        I: IntoIterator<Item = &'a GlobalId>,
+    {
+        let ids = ids.into_iter().collect_vec();
+        let mut id_bundle = CollectionIdBundle::empty();
```
```suggestion
        let mut id_bundle = CollectionIdBundle::default();
```

ggevay replied:
Yes, thx!

mjibson commented on src/adapter/src/coord/id_bundle.rs:
```diff
@@ -26,6 +26,13 @@ pub struct CollectionIdBundle {
 }
 
 impl CollectionIdBundle {
+    pub fn empty() -> Self {
```
This can get removed, use default instead.

ggevay replied:
done

mjibson commented on src/adapter/src/coord/command_handler.rs:
```diff
@@ -821,11 +840,42 @@ impl Coordinator {
+                self.acquire_read_holds_auto_cleanup(
+                    session,
+                    timestamp,
+                    &self.sufficient_collections_all_clusters(resolved_ids.0.iter()),
```
Why is it correct to use all clusters here instead of the active cluster?

ggevay replied:
Well, I didn't want to move or duplicate the cluster selection logic here, which currently happens [later](https://github.com/MaterializeInc/materialize/blob/e4cce43ce8b4fe54c57fe8edf3db94d929952cad/src/sql/src/plan/statement/ddl.rs#L2140-L2143).

I think it shouldn't cause any problems if we have read holds on a few more objects on some other clusters for a short while. But if this is too scary from a workload isolation perspective (e.g., optimizing the MV might take forever, at least for now that we don't have a timeout on optimizer calls), then maybe I can move or duplicate the cluster selection logic here, and then just acquire head holds from the relevant cluster. What do you think?

mjibson replied:
I think it's worth doing that work to avoid possible isolation problems.

ggevay replied:
Ok, then I'll factor out that logic into a function, and call it also here.

ggevay replied:
Unfortunately, this is turning out more painful than expected, because `resolve_cluster` is a method of `StatementContext`, and we don't yet have a `StatementContext` in the purification.

It uses the `StatementContext` just to access the catalog, so I can probably hack this somehow, but I'm not really sure whether it's worth it at this point. E.g., maybe I can make a `resolve_cluster` variation that takes a `&'a dyn SessionCatalog` instead of a `StatementContext`.

mjibson replied:
> maybe I can make a `resolve_cluster` variation that takes a `&'a dyn SessionCatalog`

Seems right to me. The StatementContext impl is just a wrapper around that anyway.

ggevay replied:
Ok, done!

ggevay commented:
There is some new, rather scary-looking error in the test that I added in the slt. I'll investigate tomorrow.

ggevay commented:
Actually, it turned out that I can trigger the above-mentioned error also on main. If I simply add
```
statement ok
select mz_unsafe.mz_sleep(3);
```
to the end of `materialized_views.slt` then I get this soft_panic:
```
frontier advancements after the `until` should be suppressed
```
I'll have to fix this now before merging this PR.

ggevay commented:
(To unblock the slt run, I (possibly temporarily) pushed a commit that flips the `persist_txn_tables_cli` switch back in the slt runner config. Slack discussion [here](https://materializeinc.slack.com/archives/C01CFKM1QRF/p1705490015764119).)

ggevay commented:
I've addressed all comments. Thanks for the review!

There is currently a commit that removes a `soft_panic`, I'll revert this after Dan's PR (https://github.com/MaterializeInc/materialize/pull/24492) is merged. (Dan's PR will solve the soft_panic for this PR, but we might still decide to remove the soft_panic later, depending of the outcome of [this discussion](https://materializeinc.slack.com/archives/CMNT2QSBZ/p1705509940786839). However, even if we ultimately decide to remove it, that should be done in a separate PR.)

mjibson commented on src/adapter/src/coord/command_handler.rs:
```diff
@@ -835,11 +834,15 @@ impl Coordinator {
-                    timestamp,
-                    &self.sufficient_collections_all_clusters(resolved_ids.0.iter().collect_vec()),
-                );
+                let owned_catalog = self.owned_catalog();
```
This can use `.catalog()` since it's executing in thread.

aalexandrov commented on src/compute/src/sink/refresh.rs:
```diff
@@ -115,7 +115,7 @@ where
                             // the last refresh, but not empty, which would be a bug somewhere in
                             // the `until` handling.
-                            soft_panic_or_log!("frontier advancements after the `until` should be suppressed");
+                            /////////////////soft_panic_or_log!("frontier advancements after the `until` should be suppressed");
```
Why is this commented out?

ggevay replied:
See here: https://github.com/MaterializeInc/materialize/pull/24443#issuecomment-1896468096

ggevay replied:
I've remove this soft_panic in an other PR: https://github.com/MaterializeInc/materialize/pull/24582
